File: 24. Swap Nodes in Pairs Medium/solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def swapPairs(head: ListNode) -> ListNode:
    dummy = ListNode(0)
    dummy.next = head
    prev = dummy

    while prev.next and prev.next.next:
        first, second = prev.next, prev.next.next
        prev.next, first.next, second.next = second, second.next, first

        prev = first

    return dummy.next

def createLinkedList(lst: list) -> ListNode:
    dummy = ListNode(0)
    current = dummy
    for val in lst:
        current.next = ListNode(val)
        current = current.next
    return dummy.next

def linkedListToList(head: ListNode) -> list:
    lst = []
    current = head
    while current:
        lst.append(current.val)
        current = current.next
    return lst

head1 = createLinkedList([1, 2, 3, 4])
print(linkedListToList(swapPairs(head1)))

head2 = createLinkedList([])
print(linkedListToList(swapPairs(head2)))

head3 = createLinkedList([1])
print(linkedListToList(swapPairs(head3)))

File: 24. Swap Nodes in Pairs Medium/solution.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {
	head1 := createLinkedList([]int{1, 2, 3, 4})
	fmt.Println(linkedListToList(swapPairs(head1)))

	head2 := createLinkedList([]int{})
	fmt.Println(linkedListToList(swapPairs(head2)))

	head3 := createLinkedList([]int{1})
	fmt.Println(linkedListToList(swapPairs(head3)))
}

func swapPairs(head *ListNode) *ListNode {
	dummy := &ListNode{}
	dummy.Next = head
	prev := dummy

	for prev.Next != nil && prev.Next.Next != nil {
		first := prev.Next
		second := prev.Next.Next

		prev.Next = second
		first.Next = second.Next
		second.Next = first

		prev = first
	}

	return dummy.Next
}

func createLinkedList(lst []int) *ListNode {
	dummy := &ListNode{}
	current := dummy
	for _, val := range lst {
		current.Next = &ListNode{Val: val}
		current = current.Next
	}
	return dummy.Next
}

func linkedListToList(head *ListNode) []int {
	lst := []int{}
	current := head
	for current != nil {
		lst = append(lst, current.Val)
		current = current.Next
	}
	return lst
}
